### PR TITLE
NSV Holopad Port

### PR DESCRIPTION
**Pull request by fighterslam**
## About The Pull Request
Ports the NSV admin holopad code - allowing admins to appear as special centcom holograms on-station. 
I also plan on porting some other code from /tg/ to modernize holopads a bit, clean up the UI, etc. 

Credit of the sprites/sounds/code goes to KMC, I just ported it from https://github.com/BeeStation/NSV13/pull/594

## Why It's Good For The Game
Allows admins to do more event-wise, and interact with players better!

## Changelog
:cl:fighterslam, KMC
add: Adds the admin-holopad interaction feature from our downstream, NSV13.
/:cl:

### CODE REVIEW

File: code/game/machinery/hologram.dm [hunk 1]
```diff
@@ -704,3 +704,80 @@ For the other part of the code, check silicon say.dm. Particularly robot talk.*/
 
 #undef HOLOPAD_PASSIVE_POWER_USAGE
 #undef HOLOGRAM_POWER_USAGE
+
+/obj/machinery/holopad/proc/create_admin_hologram(client/C)
```
Comment (ike709) [inline code/game/machinery/hologram.dm:708]: This entire proc is very expensive.

File: code/game/machinery/hologram.dm [hunk 2]
```diff
@@ -704,3 +704,80 @@ For the other part of the code, check silicon say.dm. Particularly robot talk.*/
...
+	var/obj/machinery/holopad/source = null
+
+/mob/living/simple_animal/admin_holopad/Initialize()
+	. = ..()
+	source = locate(/obj/machinery/holopad) in get_turf(src)
+	current_beam = new(src,source,time=INFINITY,maxdistance=INFINITY, beam_icon_state="hologram",btype=/obj/effect/ebeam)
+	INVOKE_ASYNC(current_beam, /datum/beam.proc/Start)
+	RegisterSignal(src, COMSIG_MOVABLE_MOVED, .proc/check_distance)
+
+/mob/living/simple_animal/admin_holopad/proc/check_distance()
```
Comment (ike709) [inline code/game/machinery/hologram.dm:759]: Missing `SIGNAL_HANDLER`

File: code/game/machinery/hologram.dm [hunk 3]
```diff
@@ -704,3 +704,80 @@ For the other part of the code, check silicon say.dm. Particularly robot talk.*/
...
+		to_chat(src, "<span class='warning'>Range exceeded. Transmission terminated.</span>")
+		qdel(src)
+
+/mob/living/simple_animal/admin_holopad/ClickOn(target)
+	. = ..()
+	if(alert("End communications?",src,"Yes","No") == "Yes")
+		qdel(src)
+
+/mob/living/simple_animal/admin_holopad/Destroy()
+	qdel(current_beam)
```
Comment (ike709) [inline code/game/machinery/hologram.dm:773]: That's a potential harddel right there.

Comment (Kapu1178) [reply code/game/machinery/hologram.dm:773]: Can I ask how you identify potential harddels like this?

File: code/game/machinery/hologram.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/game/machinery/hologram.dm:708]: in fairness it can only be triggered by admins

File: code/game/machinery/hologram.dm [hunk 4]
```diff
@@ -704,3 +704,80 @@ For the other part of the code, check silicon say.dm. Particularly robot talk.*/
...
+		final.Insert(I, "hologram", frame=1, dir=dir) //Then, we add this new icon_state and direction to the icon we're generating. This is then cleanly applied to the dummy mob to give it its appearance.
+	qdel(D)
+	say("Incoming priority one transmission from Central Command.")
+	playsound(src, 'sound/effects/admin_holopad_activate.ogg', 100, FALSE)
+	sleep(3.6 SECONDS)
+	var/mob/Hologram = new /mob/living/simple_animal/admin_holopad(get_turf(src))
+	Hologram.alpha = 0
+	Hologram.name = (new_name) ? new_name : Hologram.name
+	Hologram.icon = final
+	Hologram.grant_all_languages(grant_omnitongue=TRUE) //This is an admin thing, so this makes sense to me.
```
Comment (PowerfulBacon) [inline code/game/machinery/hologram.dm:734]: Can't admins turn on omnitongue if they want it anyway?

File: code/game/machinery/hologram.dm [hunk 5]
```diff
@@ -704,3 +704,80 @@ For the other part of the code, check silicon say.dm. Particularly robot talk.*/
...
+		randomize_human(D)
+		new_name = input(C, "Select a name for your communications hologram (leave blank to just be called hologram)", "Robust hologram creator") as text
+	C.cmd_admin_dress(D)
+	for(var/dir in list(2, 1, 4, 8)) //A little bit hacky, but does the trick. BYOND's icon format starts with a south facing dir, while cardinals starts with NORTH.
+		CHECK_TICK
+		D.setDir(dir) //Set the dir of the mannequin to face the next direction
+		COMPILE_OVERLAYS(D) //Prepare it for an image capture
+		var/icon/I = icon(getFlatIcon(D), frame = 1) //And finally clone the appearance of our new dummy character.
+		final.Insert(I, "hologram", frame=1, dir=dir) //Then, we add this new icon_state and direction to the icon we're generating. This is then cleanly applied to the dummy mob to give it its appearance.
+	qdel(D)
```
Comment (PowerfulBacon) [inline code/game/machinery/hologram.dm:726]: I hate this making a human and then immediately deleting them, but I guess thats how its done anyway

File: code/game/machinery/hologram.dm (hunk 3 above)
Comment (PowerfulBacon) [reply code/game/machinery/hologram.dm:773]: Anything referenced by something else when its deleted hard dels. This includes if its in vis_contents and stuff like that.
If you have atom A referenced by B, when A is deleted it needs to unreference itself in B in the destroy proc.

Comment (fighterslam) [reply code/game/machinery/hologram.dm:773]: oooookay, how do I fix that? Why is that a problem? I'm just porting this, but I'm still very new to this whole coding thing and don't understand the intricacies of everything yet.

**Pull request closed without merging**